elastica/joint.py

```diff
@@ -222,51 +228,65 @@ def __init__(self, k, nu, kt):
+            If True, the initial rotation offset between the two systems is recorded
+            and enforced throughout the entire simulation.
         """
         super().__init__(k, nu)
         # additional in-plane constraint through restoring torque
         # stiffness of the restoring constraint -- tuned empirically
         self.kt = kt
+        self.nut = nut
+
+        # if a static rotation offset between the two systems should not be enforced,
+        # we set the relative rotation to the identity matrix. Otherwise
+        if use_static_rotation:
+            self.static_rotation = None
+        else:
+            self.static_rotation = np.eye(3)
 
     # Apply force is same as free joint
     def apply_forces(self, rod_one, index_one, rod_two, index_two):
         return super().apply_forces(rod_one, index_one, rod_two, index_two)
 
-    def apply_torques(self, rod_one, index_one, rod_two, index_two):
-        # current direction of the first element of link two
-        # also NOTE: - rod two is fixed at first element
-        link_direction = (
-            rod_two.position_collection[..., index_two + 1]
-            - rod_two.position_collection[..., index_two]
-        )
+    def apply_torques(self, system_one, index_one, system_two, index_two):
+        # collect directors of systems one and two
+        # note that systems can be either rods or rigid bodies
+        system_one_director = system_one.director_collection[..., index_one]
+        system_two_director = system_two.director_collection[..., index_two]
 
-        # To constrain the orientation of link two, the second node of link two should align with
-        # the direction of link one. Thus, we compute the desired position of the second node of link two
-        # as check1, and the current position of the second node of link two as check2. Check1 and check2
-        # should overlap.
+        if self.static_rotation is None:
+            # this if clause should be active during the first timestep for the case use_static_rotation==True
+            self.static_rotation = system_one_director @ system_two_director.T
 
-        tgt_destination = (
-            rod_one.position_collection[..., index_one]
-            + rod_two.rest_lengths[index_two] * rod_one.tangents[..., index_one]
-        )  # dl of rod 2 can be different than rod 1 so use rest length of rod 2
+        # relative rotation matrix from system 1 to system 2: C_12 = C_1I @ C_I2 where I denotes the inertial frame
+        rel_rot = system_one_director @ system_two_director.T
+        # C_22* = C_21 @ C_12* where C_12* is the desired rotation between systems one and two
+        error_rot = rel_rot.T @ self.static_rotation
 
-        curr_destination = rod_two.position_collection[
-            ..., index_two + 1
-        ]  # second element of rod2
+        # compute rotation vectors based on C_22*
+        rot_vec = Rotation.from_matrix(error_rot).as_rotvec()
```

Comment:
You can use `_get_rotation_matrix` instead to be consistent, and for future `numba` wrapping.

Reply:
@skim0119 `_get_rotation_matrix` is not suitable as it computes the rotation matrix. I need to compute the rotation vector based on a rotation matrix. 

Do you think it is advantageous to implement this with numba? I am asking, because we do not need any batched processing in this instance, but only need to compute the rotation vector for one rotation matrix.

Reply:
We have `_inv_rotate` for Rodrigues formula as well. You can use this.

We probably don't have much speed up with `numba` in this case, but we are trying to keep the code consistent for later. We are constantly checking if there are new traits available for optimization (`numba` for past example), and we want to keep dependencies as consistent as possible, especially for the critical kernels. It is why we are a little hesitant in utilizing high-level packages, especially during the stepping.

Reply:
Maybe `inv_rotate` is what you may need @mstoelzle?

Reply:
@skim0119 I am confused about the implementation of `_inv_rotate`. Could you please expand a bit what exactly it does? 

Particularly, I have a question concerning the shape of the return array. I would expect that similarly to other functions for example in `_linalg.py`, the functions expects inputs in batch format (with the `n` dimension last). However, the dimension of the last dim is calculated as `n-1` in line 97. This is also inconsistent with the docstring, which says that the return has a shape of `(dim, n)`. But in reality, it is `(dim, n-1)`.
```py
@njit(cache=True)
def _inv_rotate(director_collection):
    """
    Calculated rate of change using Rodrigues' formula

    Parameters
    ----------
    director_collection : The collection of frames/directors at every element,
    numpy.ndarray of shape (dim, dim, n)

    Returns
    -------
    vector_collection : The collection of axes around which the body rotates
    numpy.ndarray of shape (dim, n)

    Note
    ----
    TODO: Benchmark missing

    """
    blocksize = director_collection.shape[2] - 1
    vector_collection = np.empty((3, blocksize))
```

Reply:
@skim0119 Also, when I compare the outputs of Scipy `Rotation.from_matrix(dev_rot).as_rotvec()` and `_inv_rotate(dev_rot.reshape(3, 3, 1)).squeeze()` for the same rotation matrix `dev_rot.reshape `, the outputs are quite different:
```
dev_rot:  [[ 1.00000000e+00 -2.98624524e-07 -1.39620560e-10]
 [ 2.98624524e-07  1.00000000e+00  2.75057444e-07]
 [ 1.39537193e-10 -2.75057444e-07  1.00000000e+00]]
theta 4.060023871629602e-07 n [-6.77477405e-01 -3.43788316e-04  7.35524059e-01]
Scipy Rotation.from_matrix(dev_rot).as_rotvec(): [-2.75057444e-07 -1.39578877e-10  2.98624524e-07]
_inv_rotate(dev_rot.reshape(3, 3, 1)).squeeze(): [-1.20919958e+00 -1.05479264e-06 -1.20919958e+00]

```

Reply:
Hi, @mstoelzle good catch regarding the documentation @armantekinalp take note, as you see `inv_rotate` actually takes in a collection of n directors and returns n - 1 rotation vectors, where the vectors correspond to the relative rotation vectors between consecutive directors in the collection. A possible way of using inv_rotate to get the rotation vector in this scenario is:
```
_inv_rotate(np.dstack([rel_rot, self.rest_rotation_matrix]))
```
Check once implemented if the signs are consistent.

Reply:
@bhosale2 @armantekinalp @skim0119 Yes, I think it would be a good idea to improve the documentation on `_inv_rotate` and explaining its purpose, input & output a bit better. Possibly, it would also be nice to change the name of the method to something more intuitive.